### PR TITLE
submodules: Update RLS and Rustfmt

### LINKED ISSUE #70129: `rls` no longer builds after rust-lang/rust#70118
Hello, this is your friendly neighborhood mergebot.
After merging PR rust-lang/rust#70118, I observed that the tool rls no longer builds.
A follow-up PR to the repository https://github.com/rust-lang/rls is needed to fix the fallout.

cc @pietroalbini, do you think you would have time to do the follow-up work?
If so, that would be great!

And nominating for compiler team prioritization.

### LINKED ISSUE #70280: `rustfmt` no longer builds after rust-lang/rust#70275
Hello, this is your friendly neighborhood mergebot.
After merging PR rust-lang/rust#70275, I observed that the tool rustfmt no longer builds.
A follow-up PR to the repository https://github.com/rust-lang/rustfmt is needed to fix the fallout.

cc @Dylan-DPC, do you think you would have time to do the follow-up work?
If so, that would be great!

And nominating for compiler team prioritization.

### PR DESCRIPTION
**Pull request by Xanewok**
Fixes #70129.
Fixes #70280.

Regression fixed specifically with https://github.com/rust-lang/rls/commit/4a587b5fda6c2c9302562ece8c2795cd9848c553.

r? @ghost

### CODE REVIEW

**Comment by Xanewok** [comment]
Fixes only toolstate.

@bors r+ p=1

**Comment by Xanewok** [comment]
Broken again by https://github.com/rust-lang/rust/pull/69965, working on a fix.

**Comment by Xanewok** [comment]
Updated but I think we'll have to bump `rustc-ap-*` libraries:

```
error[E0605]: non-primitive cast: `std::num::NonZeroU64` as `u32`
   --> /home/xanewok/.cargo/registry/src/github.com-1ecc6299db9ec823/rustc-ap-rustc_data_structures-642.0.0/profiling.rs:312:29
    |
312 |             let thread_id = std::thread::current().id().as_u64() as u32;
    |                             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: an `as` expression can only be used to convert between primitive types. Consider using the `From` trait

error[E0605]: non-primitive cast: `std::num::NonZeroU64` as `u32`
   --> /home/xanewok/.cargo/registry/src/github.com-1ecc6299db9ec823/rustc-ap-rustc_data_structures-642.0.0/profiling.rs:471:25
    |
471 |         let thread_id = std::thread::current().id().as_u64() as u32;
    |                         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |
    = note: an `as` expression can only be used to convert between primitive types. Consider using the `From` trait

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0605`.
```

due to https://github.com/rust-lang/rust/pull/70240

**Comment by calebcartwright** [comment]
@Xanewok just an fyi that we've got the rustc-ap-* crates updated now in both racer and rustfmt. There's a couple minor changes that need to be squeezed in for rustfmt (including bumping the version of rustfmt) before updating the submods  

https://github.com/racer-rust/racer/pull/1099
https://github.com/rust-lang/rustfmt/pull/4100

**Comment by topecongiro** [comment]
@calebcartwright Thank you for your work, just published rustfmt 1.4.13 on crates.io.

**Comment by Xanewok** [comment]
@calebcartwright thank you for working on it! Just pushed a change with update to both RLS and Rustfmt and confirmed locally that this fixes both RLS and Rustfmt.

**Comment by Xanewok** [comment]
@topecongiro could you please double-check the changes to rustfmt submodule?

**Comment by Xanewok** [comment]
```
crate `rustc-ap-syntax` is missing, update `check_crate_duplicate` if it is no longer used

**Review by topecongiro — APPROVED** [review]
LGTM.

**Comment by matthiaskrgr** [comment]
@Dylan-DPC it would be great if we could land this before nightly cutoff since rustfmt has been broken for around a week now.

**Comment by matthiaskrgr** [comment]
```
[RUSTC-TIMING] rustc_ap_rustc_expand test:false 47.923
[RUSTC-TIMING] rustfmt_nightly test:false 58.577
[RUSTC-TIMING] rustfmt_format_diff test:false 5.161
[RUSTC-TIMING] rustfmt test:false 10.388
[RUSTC-TIMING] cargo_fmt test:false 11.505
[RUSTC-TIMING] git_rustfmt test:false 11.720
    Finished release [optimized] target(s) in 4m 35s
duplicate artifacts found when compiling a tool, this typically means that something was recompiled because a transitive dependency has different features activated than in a previous build:

the following dependencies are duplicated although they have the same features enabled:
the following dependencies have different features:
  crossbeam-utils 0.7.2 (registry+https://github.com/rust-lang/crates.io-index)
    `rustfmt` additionally enabled features {"nightly"} at "/Users/runner/runners/2.165.2/work/1/s/build/x86_64-apple-darwin/stage1-tools/x86_64-apple-darwin/release/deps/libcrossbeam_utils-dbe732c9e11ac54e.rlib"
    `cargo` additionally enabled features {} at "/Users/runner/runners/2.165.2/work/1/s/build/x86_64-apple-darwin/stage1-tools/x86_64-apple-darwin/release/deps/libcrossbeam_utils-6901409fd1f394ca.rlib"

to fix this you will probably want to edit the local src/tools/rustc-workspace-hack/Cargo.toml crate, as that will update the dependency graph to ensure that these crates all share the same feature set
thread 'main' panicked at 'tools should not compile multiple copies of the same crate', src/bootstrap/tool.rs:195:13
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
failed to run: /Users/runner/runners/2.165.2/work/1/s/build/bootstrap/debug/bootstrap dist
Build completed unsuccessfully in 1:07:37
== clock drift check ==
  local time: Tue Mar 31 19:03:13 UTC 2020
  network time: Tue, 31 Mar 2020 19:03:13 GMT
== end clock drift check ==

```

**Comment by mati865** [comment]
Update this line: https://github.com/rust-lang/rust/blob/75ff3110ac6d8a0259023b83fd20d7ab295f8dd6/src/tools/rustc-workspace-hack/Cargo.toml#L62

**Comment by matthiaskrgr** [comment]
rls test failure:

```
test client_use_statement_completion_doesnt_suggest_arguments ... ok
test client_workspace_symbol ... ok
test client_workspace_symbol_duplicates ... ok
thread panicked while panicking. aborting.
[2020-04-02T01:54:47Z ERROR rls::server] Can't read message
thread 'main' panicked at 'called `Result::unwrap()` on an `Err` value: Os { code: 32, kind: BrokenPipe, message: "Broken pipe" }', src/tools/rls/rls/src/server/io.rs:190:9
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
error: test failed, to rerun pass '--test client'

Caused by:
  process didn't exit successfully: `/checkout/obj/build/x86_64-unknown-linux-gnu/stage2-tools/x86_64-unknown-linux-gnu/release/deps/client-e6c2853afc83d084` (signal: 4, SIGILL: illegal instruction)


command did not execute successfully: "/checkout/obj/build/x86_64-unknown-linux-gnu/stage0/bin/cargo" "test" "--target" "x86_64-unknown-linux-gnu" "-Zbinary-dep-depinfo" "-j" "2" "--release" "--locked" "--color" "always" "--manifest-path" "/checkout/src/tools/rls/Cargo.toml" "--features" "rustc-workspace-hack/all-static" "--"
expected success, got: exit code: 101


```

**Comment by cynecx** [comment]
Would it be possible to separate the rustfmt update from this pr into a new pr and land that in the meantime?

**Comment by Xanewok** [comment]
That’s a spurious error, let’s try again:

@bors r+

**Comment by Xanewok** [comment]
```
tidy check
* 594 error codes
* highest error code: E0751
* 283 features
Checking which error codes lack tests...
Found 491 error codes
Found 0 error codes with no tests
Done!
fmt check
skip untracked path cpu-usage.csv during rustfmt invocations
skip untracked path src/doc/book/ during rustfmt invocations
skip untracked path src/doc/rust-by-example/ during rustfmt invocations
skip untracked path src/llvm-project/ during rustfmt invocations
[TIMING] Tidy -- 16.389
Build completed successfully in 0:00:40
+ /scripts/validate-toolstate.sh
Cloning into 'rust-toolstate'...
Traceback (most recent call last):
  File "../../src/tools/publish_toolstate.py", line 305, in <module>
    cur_datetime
  File "../../src/tools/publish_toolstate.py", line 205, in update_latest
    maintainers = ' '.join('@'+name for name in MAINTAINERS[tool])
KeyError: u'rustc-dev-guide'
```
Did this change in the meantime? I'll rebase just to be sure

**Comment by Xanewok** [comment]
check passes locally, so I'll just

@bors r+

**Comment by Xanewok** [comment]
The mingw-check job failed again, which leads me to believe there's a bad config for `rustc-dev-guide` in terms of toolstate checking? I'll take a look

Hopefully @bors retry will yield for now

**Comment by Dylan-DPC** [comment]
i think there's a CI issue, investigating 

@bors retry

**Comment by Xanewok** [comment]
@Dylan-DPC I believe this needs 4387a8b96e4ae5b0c0450b284574d8307b4d5685 first (cc https://github.com/rust-lang/rust/pull/69898 and https://github.com/rust-lang-nursery/rust-toolstate/pull/16)

**Comment by Dylan-DPC** [comment]
@Xanewok maybe, but i'm doubting there's a CI problem as well because other PRs are failing. (and i forgot this has high priority so thought retry would yield xD)

@bors p=1 retry

**Comment by Xanewok** [comment]
Added 4387a8b96e4ae5b0c0450b284574d8307b4d5685, let's see what `mingw-check` says about that

File: src/tools/publish_toolstate.py [hunk 1]
```diff
@@ -202,7 +202,7 @@ def update_latest(
                 old = status[os]
                 new = s.get(tool, old)
                 status[os] = new
-                maintainers = ' '.join('@'+name for name in MAINTAINERS[tool])
+                maintainers = ' '.join('@'+name for name in MAINTAINERS.get(tool, ())
```
Comment (JohnTitor) [inline src/tools/publish_toolstate.py:205]:
```suggestion
                maintainers = ' '.join('@'+name for name in MAINTAINERS.get(tool, ()))
```

Comment (JohnTitor) [reply src/tools/publish_toolstate.py:205]: I think this should fix another failure, hopefully..

Comment (JohnTitor) [reply src/tools/publish_toolstate.py:205]: Oh, original PR has the same fix :)
@Xanewok You will continue the work here, or on the original?

Comment (Xanewok) [reply src/tools/publish_toolstate.py:205]: @JohnTitor thanks for spotting this as well! I'll continue there to land this asap for everyone else

**Comment by Xanewok** [comment]
@bors p=50

Let's land this only after https://github.com/rust-lang/rust/pull/69898, please (which should unbreak the CI first)

**Comment by Dylan-DPC** [comment]
@Xanewok can you ping me somewhere? :P

**Comment by Xanewok** [comment]
Let's queue this up again: @bors r+

**Comment by Xanewok** [comment]
Just to be clear, the `mingw-check` is failing because this isn't rebased on top of #69898 but homu/bors should be unaffected since they're testing on a merge with master.

**Comment by Dylan-DPC** [comment]
@Xanewok can you rebase this?

**Comment by Xanewok** [comment]
@Dylan-DPC if this won't automatically stop the current homu run then I can rebase this, sure.

**Comment by Xanewok** [comment]
I think that was the case unfortunately

@bors r+

**Pull request merged**